### DISCUSSION TITLE
train from graphs : labels.json

### DISCUSSION THREAD

**Discussion by chtchelkatchev** [CederGroupHub/chgnet#177] [Q&A]
Dear CHGNet users! Training Chgnet using graphs (via Graph Data) implies the presence of a labels.json file. As far as I understand, this file contains energies, forces,... for each graph. How to create graphs from VASP data is well explained in fine_tuning.ipynb. It is unclear how to create labels.json. I tried to use as labels.json a JSON file created when converting vasp data, but this was wrong, of cause. File make_graphs.py I also have investigated, but everything there is optimized for data downloaded from the MaterialsProject. Can you please add instructions (or an example) to the fine-tuning.ipynb on how to create a labels.json file that is compatible with the graphs generated from the vasp dataset? Thank You in advance.

**Comment by janosh** [comment]
can you be more specific about 

> `make_graphs.py` I also have investigated, but everything there is optimized for data downloaded from the MaterialsProject

what issues are you hitting?

`make_graphs.py` takes a [`StructureData`](https://github.com/CederGroupHub/chgnet/blob/b819ef55f6dd87cc1e2b450c41cd47e5b06544e8/chgnet/data/dataset.py#L28) object which in turn takes any `Sequence` of structures along with energy, force, stress and magmom labels.

```py
struct_data = StructureData(
    structures=structures,
    energies=your_energies,
    forces=your_forces,
    stresses=your_stresses,
    magmoms=your_magmoms,
)
make_graphs(struct_data)
```

should give you the `labels.json` file

**Comment by chtchelkatchev** [comment]
I get then:

Traceback (most recent call last):
  File "vasp_to_p.py", line 80, in <module>
    make_graphs(data=struct_data,graph_dir="graph_dir")
  File "make_graphs.py", line 43, in make_graphs
    for idx, (mp_id, graph_id) in enumerate(data.keys):
TypeError: cannot unpack non-iterable numpy.int64 object

**Reply by janosh** [reply]
what's your `data.keys`? it helps if you share a code snippet and minimal data to reproduce your error

**Comment by chtchelkatchev** [comment]
[CHGNet_structures.p.gz](https://github.com/user-attachments/files/16120334/CHGNet_structures.p.gz)

**Reply by chtchelkatchev** [reply]
This is file with data

**Comment by chtchelkatchev** [comment]
It seems, there is a problem in make_graphs.py related to mp_id variable that is not present in data converted from vasp. Problems start from  "for idx, (mp_id, graph_id) in enumerate(data.keys):". In fact, enumerate(data.keys)=[0 1 ...] There is no space for mp_id...

**Comment by haidi-ustc** [comment]
@chtchelkatchev hi, did you sovle this problem ?  Could you please provide more details on how to specify  graph files to use during training?   

```
from chgnet.model import CHGNet
from chgnet.trainer import Trainer

from chgnet.data.dataset import StructureData, get_train_val_test_loader
from monty.serialization import dumpfn,loadfn

data = loadfn("chgnet_f10v20.json")

dataset = StructureData(
    structures=data['st'],
    energies=data["e"],
    forces=data["f"],
    stresses=data["s"],  # can be None
    magmoms=None,
)
train_loader, val_loader, test_loader = get_train_val_test_loader(
    dataset, batch_size=500, train_ratio=0.9, val_ratio=0.05)

model = CHGNet(
    atom_fea_dim=64,
    bond_fea_dim=64,
    angle_fea_dim=64,
    composition_model="MPtrj",
    num_radial=31,
    num_angular=31,
    n_conv=4,
    atom_conv_hidden_dim=64,
    update_bond=True,
    bond_conv_hidden_dim=64,
    update_angle=True,
    angle_layer_hidden_dim=0,
    conv_dropout=0,
    read_out="ave",
    gMLP_norm='layer',
    readout_norm='layer',
    mlp_hidden_dims=[64, 64, 64],
    mlp_first=True,
    is_intensive=True,
    non_linearity="silu",
    atom_graph_cutoff=6,
    bond_graph_cutoff=3,
    graph_converter_algorithm="fast",
    cutoff_coeff=8,
    learnable_rbf=True,
)

# Define Trainer
trainer = Trainer(
    model=model,
    targets='efsm',
    energy_loss_ratio=1,
    force_loss_ratio=1,
    stress_loss_ratio=0.1,
    mag_loss_ratio=0.1,
    optimizer='Adam',
    weight_decay=0,
    scheduler='CosLR',
    scheduler_params={'decay_fraction': 0.5e-2},
    criterion='Huber',
    delta=0.1,
    epochs=1000,
    starting_epoch=0,
    learning_rate=5e-3,
    use_device='cuda',
    print_freq=10
)
```
does it mean the above code will auto search and load the labels.json  ile that contains graph info ?

**Reply by haidi-ustc** [reply]
this problem has been solved. pls refer to 
https://gitee.com/haidi-hfut/useful-script/blob/master/chgnet/make_graphs.py
https://gitee.com/haidi-hfut/useful-script/blob/master/chgnet/train_with_graph.py
